concrete/src/Http/ResponseFactory.php

```diff
@@ -0,0 +1,378 @@
+    {
+        return JsonResponse::create($data, $code, $headers);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function notFound($content, $code = Response::HTTP_NOT_FOUND, $headers = array())
+    {
+        if (strcasecmp($this->request->server->get('HTTP_X_REQUESTED_WITH', ''), 'xmlhttprequest') === 0) {
+            $loc = $this->localization;
+            $changeContext = $this->shouldChangeContext();
+            if ($changeContext) {
+                $loc->pushActiveContext('site');
+            }
+            $responseData = [
+                'error' => t('Page not found'),
+                'errors' => [t('Page not found')],
+            ];
+            if ($changeContext) {
+                $loc->popActiveContext();
+            }
+
+            return $this->json($responseData, $code, $headers);
+        }
+
+        $item = '/page_not_found';
+        $c = Page::getByPath($item);
+        if (is_object($c) && !$c->isError()) {
+            return $this->collection($c, $code, $headers);
+        }
+
+        $cnt = $this->app->make(PageForbidden::class);
+        $this->($cnt, $code, $headers);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function error($content, $code = Response::HTTP_INTERNAL_SERVER_ERROR, $headers = array())
+    {
+        return $this->create($content, $code, $headers);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function forbidden($requestUrl, $code = Response::HTTP_FORBIDDEN, $headers = array())
+    {
+        // set page for redirection after successful login
+        $this->session->set('rUri', $requestUrl);
+
+        // load page forbidden
+        $item = '/page_forbidden';
+        $c = Page::getByPath($item);
+        if (is_object($c) && !$c->isError()) {
+            return $this->collection($c, $code, $headers);
+        }
+
+        $cnt = $this->app->make(PageForbidden::class);
+        $this->controller($cnt, $code, $headers);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function redirect($to, $code = Response::HTTP_MOVED_PERMANENTLY, $headers = array())
+    {
+        return RedirectResponse::create($to, $code, $headers);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function view(View $view, $code = Response::HTTP_OK, $headers = array())
+    {
+        $changeContext = $this->shouldChangeContext();
+        if ($changeContext) {
+            $this->localization->pushActiveContext('site');
+        }
+
+        $contents = $view->render();
+        if ($changeContext) {
+            $this->localization->popActiveContext();
+        }
+
+        return $this->create($contents, $code, $headers);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function controller(Controller $controller, $code = Response::HTTP_OK, $headers = array())
+    {
+        if ($controller->isReplaced()) {
+            return $this->controller($controller->getReplacement());
```

Comment:
This is where the conflict was, we needed to reapply the response creation login when the controller is replaced by another controller.